Introduces the Content Comparison Modal for side-by-side review of version and revision changes.

robluton opened the pull request:
## Scope

What's changed:

### Introduces the Content Comparison Modal, an enhanced side-by-side interface for reviewing content changes between versions and revisions.

Visual indicators help identify and manage changes: the 'Updated' badge appears next to fields that were modified in the selected revision, while color-coded diff indicators (green on the incoming side, red on the base side) mark fields with differences between the two versions. Users can selectively choose which changed fields to apply, providing granular control over content updates.

_Updated badge next to fields changed in selected revision_
<img width="800" alt="CleanShot 2025-10-28 at 17 07 40@2x" src="https://github.com/user-attachments/assets/16b28b6c-e5d3-44ac-b916-e2b6c2dcaa3a" />

_Content editors can choose which differences to apply_
<img width="800" alt="CleanShot 2025-10-28 at 17 10 05@2x" src="https://github.com/user-attachments/assets/1667b982-221f-438d-816b-c86925c39ae3" />


**Props**
Two new props have been added to `v-form` and `form-interface` components to support comparison functionality. The `comparison` prop provides comparison context (including field differences and selection state), while the `comparison-active` boolean prop signals when components are operating within a comparison modal, enabling them to display appropriate visual indicators and interactive controls.


## Potential Risks / Drawbacks

- Some unused translations were removed
- The comparison modal replaces the revision drawer and now appears everywhere the drawer did—across regular (non‑system) collections, but also system collections like files, users, and policies.

## Tested Scenarios

- Version Promotion
- Opening a revision from the sidebar of the main item or a version

## Review Notes / Questions

- Test by promoting a version and applying revisions to custom and main versions


## Checklist

- [x] Added or updated tests
- [x] Documentation PR created [here](https://github.com/directus/docs/pull/496)

robluton committed:
basic layout

robluton committed:
styling

robluton committed:
add metadata to version heading

robluton committed:
rename version-promote-drawer to comparison-modal

robluton committed:
styling and fields edited count

robluton committed:
Select all checkbox

robluton committed:
remove css no longer being used

robluton committed:
add border below the version headers

robluton committed:
rework css and class naming

robluton committed:
extract header component

robluton committed:
add checkbox to interfaces

robluton committed:
only show checkboxes next to fields in the version

robluton committed:
create a diff between version and main, as a version can include fields with values that match those on main

robluton committed:
simplify diff

robluton committed:
green and red color indicators

robluton committed:
refactor to pass down object instead of several props

robluton committed:
responsive styling

robluton committed:
Merge branch 'main' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
changeset

robluton committed:
lint changes

robluton committed:
remove redundant types and unneeded defaults

robluton committed:
move computed vars and methods into composable

robluton committed:
incorporate revisions into content comparison modal

robluton committed:
get checkboxes to work with revisions

robluton committed:
refactor

robluton committed:
separate normalizer into its own util

robluton committed:
refactor to use new normalizer

robluton committed:
delta selector

robluton committed:
consolidate and cleanup

robluton committed:
simplify tests

robluton committed:
add versions to selectableDeltas for consistency

robluton committed:
more refactoring - normalize more data to allow the modal to get it from the same places

robluton committed:
the latest tag should only show up if the base is showing the latest state of the current version

robluton committed:
tests and small refactor

robluton committed:
remove dead code from earlier refactor

robluton committed:
fix bug with latest tag for version

robluton committed:
use consistent terminology for comparison sides

robluton committed:
cleanup

robluton committed:
add translations

formfcw commented on app/src/modules/content/components/comparison-modal.vue:
Let’s improve the responsive breakpoints in a way, to show the side by side comparison without horizontal scrollbar as long as possible. This means, as soon as a half width field gets smaller then 262px (or the row gets smaller than 556px), display every half width field as full width — happens around 1330px screen width. For smaller screens with a horizontal scrollbar, let’s set the breakpoint as soon as the field width gets smaller then 262px — breakpoint around 706px.

formfcw reviewed (changes requested):
Thank you @robluton :rocket: Looks great so far!

Besides the code comments, I noticed the following:

- [x] It takes a while for the modal to appear. Is there a way to improve this, such as adding a skeleton-loader to the modal?
- [x] Version comparison: When changing values in main after the version has been created, it shows different values but there is no indication of it. This should work like it did before.

robluton commented on app/src/components/v-form/types.ts:
```diff
@@ -6,3 +6,11 @@ export type FormField = DeepPartial<Field> & {
+
+export interface ComparisonContext {
+	mode: boolean;
+	side?: 'main' | 'current';
```
I agree that the terminology you are suggesting is more clear. I was using different terms myself until I decided to mirror the terms used by the version compare endpoint. I think the consistency gained here is something we should consider. Wdyt?

robluton committed:
put toggleComparison inside comparison context object

robluton committed:
remove data attrs

robluton committed:
cleanup

robluton committed:
copmarison-specific breakpoints

robluton committed:
Update .changeset/free-trains-clap.md

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
checkbox styles

robluton committed:
loading state

formfcw commented on app/src/styles/mixins/_form-grid.scss:
```diff
@@ -1,5 +1,6 @@
 @mixin form-grid {
 	display: grid;
+	align-items: start;
```
Let's think through this change again. Could it break something?

formfcw commented on app/src/modules/content/components/comparison-header.vue:
```diff
@@ -0,0 +1,302 @@
+				</div>
+			</template>
+			<template v-else>
+				<div v-if="showDeltaDropdown" class="delta-dropdown">
```
Let’s rename `delta` to `revision` since the dropdown will always display revision items.

formfcw commented on app/src/modules/content/components/comparison-header.vue:
```diff
@@ -0,0 +1,302 @@
+			</template>
+			<template v-else>
+				<div v-if="showDeltaDropdown" class="delta-dropdown">
+					<v-menu show-arrow>
```
Let’s let the menu appear attached to the button.

```suggestion
					<v-menu attached>
```

formfcw commented on app/src/modules/content/components/comparison-header.vue:
```diff
@@ -0,0 +1,302 @@
+								<v-list-item-content>
+									<div class="delta-option-content">
+										<div class="delta-option-date-time">{{ option.text.split(' - ')[0] }}</div>
+										<div class="delta-option-user-info">{{ option.text.split(' - ')[1] }}</div>
```
Let’s use the same representation as above.

```suggestion
										<div class="delta-option-user-info">{{ t('edited_by') }} {{ option.text.split(' - ')[1] }}</div>
```

formfcw commented on app/src/modules/content/components/comparison-header.vue:
```diff
@@ -0,0 +1,302 @@
+						</template>
+						<v-list>
+							<v-list-item
+								v-for="option in deltaOptions"
```
Let's consider adding pagination to the dropdown menu when there are a large number of revisions. What do you think?

formfcw reviewed (changes requested):
Thank you for the updates.

What I noticed with Revisions:
- [x] The two sides appear to be reversed. The "current" state of the item should always appear on the left side and the "incoming" state should always appear on the right. I assume that’s why nothing happens when clicking "Revert" — see also the comment about `current | incoming`.
- [x] The max updated fields indicatior shows "2 fields" when it should show "1 field". The number also differs from the listing in the sidebar.
- [x] The date is always the same on both sides. When I change the date on one side, it changes on the other side too.
- [x] Previously the first entry of the revision sidebar opened the changes, that with your implementation open when clicking the second entry. Let’s stick to the old behavior. Example: I have one revision entry with "1 Field" updated. Then I update 30 fields and a new Revision entry "Updated 30 fields" appears in the sidebar. When I click this entry, I expect to be able to revert the changes to the 30 fields, but with your implementation it shows the current state instead.

- [x] Let’s also remove the `version-promote-field` component, which was only used in the `version-promote-drawer` component.

robluton committed:
change terminology for left and right sides

formfcw reviewed (changes requested):
A few things about the design

robluton committed:
change revision revert to promote

robluton committed:
remove legacy revisions code

robluton committed:
use subdued foreground color for unchecked checkbox

robluton committed:
Update app/src/modules/content/components/comparison-modal.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-header.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-header.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-header.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-header.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-header.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
updates

robluton committed:
updates

robluton committed:
Merge branch 'robl/cms-1246-content-comparison-modal-core' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
keep full width fields until 1330px

robluton committed:
updates

robluton commented on app/src/modules/content/components/comparison-modal.vue:
```diff
@@ -0,0 +1,610 @@
+			grid-column: start / full;
+		}
+	}
+}
```
had to go about this slightly differently, but should be the correct outcome.

formfcw commented on app/src/modules/content/components/comparison-header.vue:
```diff
@@ -0,0 +1,302 @@
+				<div v-if="showDeltaDropdown" class="delta-dropdown">
+					<v-menu show-arrow>
+						<template #activator="{ toggle }">
+							<div class="delta-selection" @click="toggle">
```
- [ ] As discussed internally, let’s also use the default border-width (2px instead of 1px)

robluton committed:
more feedback updates

robluton committed:
feedback updates

robluton committed:
remove latest tag feature

robluton committed:
change sides to base and incoming to avoid confusion with version comparison api response

robluton committed:
adjust tests

robluton committed:
style tweaks

robluton committed:
change title of left side back to Main

robluton committed:
pull latest

robluton committed:
use the term Latest in place of the date on left side of comparison

robluton committed:
calculate updated fields for revisions list

robluton committed:
label first revision in drawer as Latest

robluton committed:
Show hidden fields in comparison view (#25878)

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

formfcw commented on app/src/modules/content/components/comparison-modal.vue:
```diff
@@ -0,0 +1,610 @@
+		flex: 0 0 auto;
+		justify-content: space-between;
+		padding-inline: var(--comparison-modal-padding-x);
+		padding-block: var(--comparison-modal-padding-y);
```
For the footer it is 18px. The reason is that the button height is 44px. 18 + 44 + 18 = 80 -> The height without border should be 80px for both header and footer. No worries, I will handle it with a separate PR.

<img width="1432" height="608" alt="18" src="https://github.com/user-attachments/assets/a3013f9f-515e-4c05-b045-3f90ac5f8459" />

formfcw reviewed (changes requested):
Thank you @robluton!

- [x] Please have a look at the unresolved change requests from above:
    - https://github.com/directus/directus/pull/25782#discussion_r2352100865
    - https://github.com/directus/directus/pull/25782#discussion_r2355757015
    - and the Todo here: https://github.com/directus/directus/pull/25782#pullrequestreview-3228135488
- [x] Let’s remove the unused component app/src/views/private/components/revisions-drawer-updates-change.vue

robluton committed:
Apply suggestions from code review

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Update app/src/modules/content/components/comparison-modal.vue

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
use new translations

formfcw committed:
Refine comparison component styles and structure (#25897)

* apply margin only to comparison checkbox

* update comparison padding values according to the design

* remove unnecessary wrapper from the title

* simplify meta-content structure and remove dropdown class

* ensure meta selection text always aligns left/start

* add hover style for dropdown

* adjust focus ring offset of dropdown button

* adjust dropdown option styling

* reuse delta option time and user extraction logic

* simplify vertical divider

robluton committed:
Use "Latest" instead of "Created"

robluton committed:
Merge branch 'robl/cms-1246-content-comparison-modal-core' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
fix issue causing field count differences

robluton committed:
filter out read-only fields after api call

robluton committed:
add debug button to help investigate any issues with the diff

robluton committed:
do not allow fields that do not exist in the collection itself

robluton committed:
fix tests

robluton committed:
remove more legacy code

robluton committed:
rename revisions-drawer-detail to revisions-sidebar-detail

robluton committed:
Prevent empty fields when trying to promote a version that has not diverged from main

formfcw committed:
Refine comparison style (#25904)

* simplify grid adjustments for comparison forms

* improve field label responsiveness

* add intermediate breakpoint for the column width

* update header height according to design

* fix responsiveness of comparison header

* adjust padding for meta text in comparison header according to design

* improve responsiveness and text-overflow for comparison header

* remove extra flex-wrap property

---------

Co-authored-by: Rob Luton <rob.luton@gmail.com>

robluton committed:
change terminology from "udpated" to "differences"

robluton committed:
do not allow click event on latest revision

robluton committed:
do not include latest revision in comparison dropdown

formfcw reviewed (changes requested):
﻿﻿Thank you!

- [x] I am currently getting an error „[FORBIDDEN] You don't have permission to access this“ when entering the following routes:
  - settings > translations > detail page
  - files > detail page
  - policies > detail page
  - user > detail page

robluton committed:
Revision Updates (#25954)

Co-authored-by: formfcw <dev@formfcw.com>

formfcw reviewed (changes requested):
I have updated RELATIONAL_TYPES constant — see https://github.com/directus/directus/pull/25964.

Let’s take advantage of it — see the code comments.

First merge main into this branch and run `pnpm --filter constants --filter types build`.

robluton committed:
Merge branch 'main' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
Apply suggestions from code review

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
Merge branch 'robl/cms-1246-content-comparison-modal-core' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
Revision diff bug fixes (#25969)

formfcw commented on app/src/modules/content/comparison-utils.ts:
Let’s not skip read-only fields — see the other comment.
Therefore, please refactor all the relevant parts, such as `{ skipRelationalFields, skipReadonlyFields }` to `skipRelationalFields`.

formfcw reviewed (changes requested):
Thank you @robluton :heart:

- [x] In the revisions-comparison the M2O fields don’t show the same value on each side. Even if no change is indicated, they should show the same values. Affects the File, Image and M2O interfaces.

robluton committed:
Apply suggestions from code review

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
more updates from code review

robluton committed:
deal with revisionId and versionId type mismatch

robluton committed:
add translations

robluton committed:
Apply suggestions from code review

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

robluton committed:
skip auto date fields

robluton committed:
for revisions, use related values from main so the form is consistent

robluton committed:
use correct endpoints for system collections

robluton committed:
use new translation

robluton committed:
fix type issue

robluton committed:
ignore primary key fields and user-created and user-updated fields for comparisons

robluton committed:
fix count issue between revision list and modal

robluton committed:
skip primary key fields for version promotion

formfcw committed:
Comparison < Updates (#25991)

* simplify props usage in revisions-date-group component

* simplify field retrieval

* make array merge customizer reusable

* use existing isSystemCollection function and sort imports

* streamline comparison logic and enhance readability

* simplify currentVersion handling in comparison data normalization *

* simplify revisions handling in comparison data normalization *

* simplify versions handling in comparison data normalization *

* simplify system collection check

* replace getSystemCollectionEndpoint with existing getEndpoint function

* simplify item endpoint retrieval in useRevisions and remove isSystemCollection check

* remove unnecessary export keywords

* remove unused ComparisonSide type and related constants

* rename copyRelationalFieldsFromBaseToIncoming to copySpecialFieldsFromBaseToIncoming for clarity

* remove comment

* update css var for clarity

* use formatTitle for version key for human-readable titles

* improve footer responsiveness

formfcw committed:
simplify item fetching logic in useComparison (#25993)

formfcw committed:
Comparison < Updated Fields Indication (#26011)

* add updated fields indication

* update badge class for v-chip and adjust import order

* update translation for revision messages for clarity

formfcw committed:
improve responsiveness of field label (#26015)

formfcw committed:
Comparison < fix merge bug (#26017)

* fix merge bug

* add tests

* refactor: move getFieldsWithDifferences test into utils test file

* refactor: rename mergeMissingMainItemKeysIntoRevision to mergeMainItemKeysIntoRevision

* refactor buildRevisionComparison logic

formfcw committed:
Comparison < sidebar updates (#26042)

* squash commit robl/sidebar-updates

* (re-)use existing i18n key

* remove unused isFirstGroup prop

* remove unused computeDifferentFields function and lodash mergeWith import

* refactor: move getItemEndpoint function into useComparison and restore import

* refactor: remove unused mergeMainItemKeysIntoRevision function

* refactor: remove unnecessary differentFields from revision data structure

* improve field comparison logic and enhance revision handling

formfcw committed:
Merge branch 'main' into robl/cms-1246-content-comparison-modal-core

formfcw committed:
Comparison < Updates 27/1 (#26054)

* drill down comparisonActive prop instead of recalculating

* remove debug mode

* Ensure NormalizedItem gets correct user for the base side

* Refactor to use type PrimaryKey instead of `string | number`

* refactor to require collection parameter

* rename normalizeComparisonData to getNormalizedComparisonData for clarity

* add mainVersionMeta to ComparisonData to get activity information of latest main

* refactor user fetching logic for clarity

formfcw committed:
Comparison ← refactor, clean up and small fixes (#26063)

* use a different tooltip for the updated badge when the field auto-populates

* refactor: move data fetching logic to the modal component

* refactor: remove comparison-utils and reorganize utility functions

- Deleted the comparison-utils and moved relevant functions to new utility files and the use-comparison composable
- Refactored use-comparison composable to utilize the utility functions
- Added new field-utils for field-related checks (e.g., isPrimaryKey, isRelational)
- Added new get-revision-fields utility to handle revision field filtering logic
- Added new get-version-display-name utility for version display name formatting

* refactor: move files comparison-header.vue, comparison-modal.vue, use-comparison.ts and use-comparison.test.ts to @/views/private/components/comparison

* refactor: move type definitions to types.ts and update imports in use-comparison.ts

* refactor: rename main to base and current to incoming

* fix types by preventing renaming when newVersionKey is null

robluton committed:
Merge branch 'main' of github.com:directus/directus into robl/cms-1246-content-comparison-modal-core

robluton committed:
update changeset

robluton committed:
update changeset

formfcw committed:
Comparison ← add tests (#26076)

* add unit tests for field utility functions

* add unit tests for getRevisionFields function

* add unit tests for getVersionDisplayName function

* fix boolean checks

* add unit tests for useComparison

formfcw committed:
Comparison ← small fix, add more tests (#26078)

* clean-up

* refactor: update test descriptions for clarity on user and date fields

* fix user display and simplify logic

* simplify mock API implementation in tests and add user updated field checks

* fix visibility logic for hidden fields in form components

* ensure field labels are visible when in comparison mode

* add revisionFields to comparison data on base side

formfcw committed:
fix reactivity issue by changing currentRevision from prop to model (#26083)

formfcw reviewed (approved):
LGTM :rocket:
